RecoLocalCalo/HGCalRecProducers/plugins/HeterogeneousHGCalHEFCellPositionsConditions.cc

```diff
@@ -0,0 +1,302 @@
+          select_pointer_u_(&this->posmap_, j)[index] =
+              select_pointer_u_(cpuPos, j - this->number_position_arrays)[index];
+        } else
+          edm::LogError("HeterogeneousHGCalHEFCellPositionsConditions") << "Wrong HeterogeneousHGCalPositions type";
+      }
+    }
+  }
+}
+
+void HeterogeneousHGCalHEFCellPositionsConditions::transfer_data_to_heterogeneous_vars_(
+    const cpos::HGCalPositionsMapping* cpuPos) {
+  this->posmap_.waferSize = cpuPos->waferSize;
+  this->posmap_.sensorSeparation = cpuPos->sensorSeparation;
+  this->posmap_.firstLayer = cpuPos->firstLayer;
+  this->posmap_.lastLayer = cpuPos->lastLayer;
+  this->posmap_.waferMin = cpuPos->waferMin;
+  this->posmap_.waferMax = cpuPos->waferMax;
+  this->nelems_posmap_ = cpuPos->detid.size();
+}
+
+std::vector<size_t> HeterogeneousHGCalHEFCellPositionsConditions::calculate_memory_bytes_(
+    cpos::HGCalPositionsMapping* cpuPos) {
+  size_t npointers = cpos::types.size();
+  std::vector<size_t> sizes(npointers);
+  for (unsigned int i = 0; i < npointers; ++i) {
+    const unsigned detid_index = 4;
+    const unsigned nlayers_index = 3;
+    if (cpos::types[i] == cpos::HeterogeneousHGCalPositionsType::Float and (i == 0 or i == 1))
+      sizes[i] = select_pointer_u_(cpuPos, detid_index)
+                     .size();  //x and y position array will have the same size as the detid array
+    else if (cpos::types[i] == cpos::HeterogeneousHGCalPositionsType::Float and i == 2)
+      sizes[i] = select_pointer_i_(cpuPos, nlayers_index).size();  //z position's size is equal to the #layers
+    else if (cpos::types[i] == cpos::HeterogeneousHGCalPositionsType::Float and i > 2)
+      edm::LogError("HeterogeneousHGCalHEFCellPositionsConditions") << "Wrong HeterogeneousHGCalPositions type (Float)";
+    else if (cpos::types[i] == cpos::HeterogeneousHGCalPositionsType::Int32_t)
+      sizes[i] = select_pointer_i_(cpuPos, i - this->number_position_arrays).size();
+    else if (cpos::types[i] == cpos::HeterogeneousHGCalPositionsType::Uint32_t)
+      sizes[i] = select_pointer_u_(cpuPos, detid_index).size();
+  }
+
+  std::vector<size_t> sizes_units(npointers);
+  for (unsigned int i = 0; i < npointers; ++i) {
+    if (cpos::types[i] == cpos::HeterogeneousHGCalPositionsType::Float)
+      sizes_units[i] = sizeof(float);
+    else if (cpos::types[i] == cpos::HeterogeneousHGCalPositionsType::Int32_t)
+      sizes_units[i] = sizeof(int32_t);
+    else if (cpos::types[i] == cpos::HeterogeneousHGCalPositionsType::Uint32_t)
+      sizes_units[i] = sizeof(uint32_t);
+  }
+
+  //element by element multiplication
+  this->sizes_.resize(npointers);
+  std::transform(sizes.begin(), sizes.end(), sizes_units.begin(), this->sizes_.begin(), std::multiplies<size_t>());
+  return this->sizes_;
+}
+
+HeterogeneousHGCalHEFCellPositionsConditions::~HeterogeneousHGCalHEFCellPositionsConditions() {
+  //cudaCheck(cudaFreeHost(this->posmap_.x));
```

Comment:
this destructor is now empty - could you remove it and rely on the default one generated by the compiler ?

Reply:
Looking at it now I am no longer sure the ```cudaFreeHost()``` is not required. The guidelines recommend avoiding these calls on an event-per-event basis, but is it safe to never free the memory (or is the framework somehow responsible for it)? The conditions are used in ```RecoLocalCalo/HGCalRecProducers/plugins/HeterogeneousHGCalHEFCellPositionsFiller.h```.

Reply:
Well, no, if you *allocate* the memory, you have to free it.

By the way, I wasn't able to understand from a quick look...
  - is the memory allocated *per cell* ? or is it just one large buffer ?
  - is the memory allocated (and then freed) per event ? or per IOV ? or per job ?

Reply:
The code is currently organized in the following way:

- ```RecoLocalCalo/HGCalRecProducers/plugins/HeterogeneousHGCalHEFCellPositionsFiller.h```: fills conditions data in the CPU and ```produce```s the conditions data in the GPU (it is the ```ESProducer```, similar to [this one](https://github.com/cms-sw/cmssw/blob/dd931b1a61f3948a7c528b815bc15959d56151fa/CalibTracker/SiPixelESProducers/plugins/SiPixelROCsStatusAndMappingWrapperESProducer.cc))
- ```RecoLocalCalo/HGCalRecProducers/plugins/HeterogeneousHGCalHEFCellPositionsConditions.cc```: takes care of the CPU-GPU transfers and positions filling in the GPU
- ```CUDADataFormats/HGCal/interface/HGCConditions.h```: includes the CPU and GPU data structures which store the geometry

There is a single buffer being allocated (see ```allocate_memory_()``` in the second bullet). The allocation (and deallocation) is done once per event, which I imagine is not ideal, as the properties I am passing to the GPU will never change (unless the structure itself of HGCAL changes). What would be the recommended approach? Please note I am following the one used by the ```ESProducer``` linked above.

**Note**:
When writing the configuration file, at some point I am defining:

```python
process.task = cms.Task( process.HeterogeneousHGCalHEFCellPositionsFiller, process.HeterogeneousHGCalHEFRecHits )
```

Thus it really seems the ```ESProducer``` gets called event by event.

**Question**:
Does the linked ```ESProducer``` need for some reason to allocate the memory event by event? How is it called from a configuration file? I am guessing the requirements might be different from mine.

Reply:
Do you have some evidence the `ESProducer` actually gets called for every event? It should get called at IOV (interval of validity) boundaries of the `HeterogeneousHGCalHEFCellPositionsConditionsRecord` record, that correspond to the IOV boundaries of `IdealGeometryRecord`, that is effectively the whole job. I.e. the `ESProducer` should get called only once in the job.

Reply:
My comment was made just by looking at the code; sorry for not checking it in more detail. IMHO, it is very hard to understand exactly what ```ESProduct```s are doing due to the quite complex way they work (their interaction in the framework with ```Record```s could be simpler). For instance, one cannot simply know by heart the IOV of every record...

I thus wrote a dummy ```EDProducer``` that consumes a random collection, outputs another collection, and in between uses conditions data. The relevant bit of the configuration looks like:

```
process.Filler = cms.ESProducer("HeterogeneousHGCalHEFCellPositionsFiller")
process.Dummy = cms.EDProducer("DummyProducer",
                               Tok = cms.InputTag('HGCalUncalibRecHit','HGCHEFUncalibRecHits'))
process.task = cms.Task( process.Filler, process.Dummy )
```
The ```produce()``` of ```HeterogeneousHGCalHEFCellPositionsFiller``` gets called for the first event only. Despite calling the ```getHeterogeneousConditionsESProductAsync()``` every event, the positions are filled in the GPU only once as well. This corresponds to the behaviour described in the [documentation](https://github.com/cms-patatrack/cmssw/blob/master/HeterogeneousCore/CUDACore/README.md#cuda-esproduct). This was still true when increasing the ```numberOfThreads``` and ```numberOfStreams```. I could not play with the number of jobs, since I do not know how to change it directly and the ```patatrack-scripts/benchmark``` supresses ```cout``` output.

I personally think it would be useful for future developers to have access to the dummy HGCAL-specific "ESConsumer" I wrote. Do you think I should include it in the PR? I could store it in ```RecoLocalCalo/HGCalRecProducers/plugins/DummyESConsumer.cc```. This could be used in the near future as a template as soon as the heterogeneous version of CLUE is included.

Reply:
> For instance, one cannot simply know by heart the IOV of every record

To leading order one does not have to know the IOVs. The important part is that they change (much) less frequently than for each event, and therefore raw `cudaMalloc()` etc are not that bad (they are still suboptimal, but improvements in that area are beyond the scope of this PR).

> I personally think it would be useful for future developers to have access to the dummy HGCAL-specific "ESConsumer" I wrote.

The value of such EDModule is not immediately clear to me, but we do have bunch of test EDModules here and there. There is also a generic "ESConsumer" module in [FWCore/Modules/src/EventSetupRecordDataGetter.cc](https://github.com/cms-sw/cmssw/blob/master/FWCore/Modules/src/EventSetupRecordDataGetter.cc), but that would not (currently) trigger the host-to-device transfers.

Reply:
One can use
- PrintEventSetupContent: used to find which data products are available in which Records
- PrintEventSetupDataRetrieval: used to state when a data product has been retrieved from a Record
- Tracer service: As of last week it will report when an ESProducer or ESSource is executed